mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,318 @@
+
+    def __init__(self, discr: Discretization, quantity: str, op: str,
+                 extract_vars_for_logging, units_logging, name: str = None,
+                 axis: Optional[int] = None):
+        unit = units_logging(quantity)
+
+        if name is None:
+            name = f"{op}_{quantity}" + (str(axis) if axis is not None else "")
+
+        LogQuantity.__init__(self, name, unit)
+        StateConsumer.__init__(self, extract_vars_for_logging)
+
+        self.discr = discr
+
+        self.quantity = quantity
+        self.axis = axis
+
+        from functools import partial
+
+        if op == "min":
+            self._discr_reduction = partial(self.discr.nodal_min, "vol")
+            self.rank_aggr = min
+        elif op == "max":
+            self._discr_reduction = partial(self.discr.nodal_max, "vol")
+            self.rank_aggr = max
+        elif op == "norm":
+            self._discr_reduction = partial(self.discr.norm)
+            self.rank_aggr = max
+        else:
+            raise ValueError(f"unknown operation {op}")
+
+    @property
+    def default_aggregator(self):
+        """Rank aggregator to use."""
+        return self.rank_aggr
+
+    def __call__(self):
+        """Return the requested quantity."""
+        if self.state_vars is None:
+            return None
+
+        quantity = self.state_vars[self.quantity]
+
+        if self.axis is not None:  # momentum
+            return self._discr_reduction(quantity[self.axis])
+        else:  # mass, energy
+            return self._discr_reduction(quantity)
+
+# }}}
+
+
+# {{{ Kernel profile quantities
+
+class KernelProfile(MultiLogQuantity):
+    """Logging support for statistics of the OpenCL kernel profiling (time, \
+    num_calls, flops, bytes_accessed, footprint).
+
+    Parameters
+    ----------
+    actx
+        The array context from which to collect statistics. Must have profiling
+        enabled in the OpenCL command queue.
+
+    kernel_name
+        Name of the kernel to profile.
+    """
+
+    def __init__(self, actx: PyOpenCLArrayContext,
+                 kernel_name: str) -> None:
+        from mirgecom.profiling import PyOpenCLProfilingArrayContext
+        assert isinstance(actx, PyOpenCLProfilingArrayContext)
+
+        units = ["s", "GFlops", "1", "GByte", "GByte"]
+
+        from dataclasses import fields
+        from mirgecom.profiling import ProfileResultsForKernel
+
+        names = []
+        units = []
+
+        for f in fields(ProfileResultsForKernel):
+            names.append(f"{kernel_name}_{f.name}")
+            if f.name == "num_calls":
+                units.append("1")
+            elif f.name == "flops":
+                units.append("GFlops")
+            elif f.name == "time":
+                units.append("s")
+            elif f.name == "bytes_accessed" or f.name == "footprint_bytes":
+                units.append("GByte")
+            else:
+                raise RuntimeError(f"unknown unit for field {f.name}")
```

Comment:
This wants to be a dictionary and a list comprehension that accesses it, I think.

Reply:
Changed.